### PR TITLE
fix(drafts): publish drift guard compares a content signature, not just command count

### PR DESCRIPTION
**Pull request by youhaowei**
## Changes

- Add a deterministic SHA-256 content signature over a draft's compacted command log (`draft-log-signature.ts`), used as a second review-drift check alongside the existing `expectedCommandCount` guard.
- `DraftController.publishDraft` accepts an optional `expectedLogSignature` in `PublishDraftOptions`, enforced inside the publish transaction against the reloaded durable log (same placement/atomicity as the count guard). Kept the count check as a cheap, distinct-evidence fast path rather than folding it into the signature.
- `publishDraft` RPC (`draft-lifecycle.ts`) accepts and forwards `expectedLogSignature` (already hex text — no parsing needed).
- `draftPublishReview` (`drafts.ts`) computes and returns `logSignature` alongside the existing `commandCount`.
- Client: both review surfaces — the full publish review page and the assistant sidebar `DraftReviewPanel` — now capture `logSignature` from the review response and pass it back on publish.

### The gap this closes

The count-only guard compared `log.length !== expectedCommandCount`. `compactLog` can DROP earlier log positions (a create cancelled by a later delete collapses to nothing), so a concurrent append that triggers compaction can leave the log length unchanged from what the reviewer saw while the actual surviving content differs — same-count content drift slipped the guard undetected.

### Signature design and determinism

- Computed **server-side** in `draftPublishReview`, not client-side: the client only ever receives redacted command summaries (`{path, hasArgs, lateBoundCount}`), never raw `args`, so it has no way to compute a content signature itself. The client instead receives the server-computed `logSignature` and ferries it back verbatim as an opaque hex string.
- Hashes `[{path, args}]` in replay order — only the fields that determine what publish actually does. `id`/`compactionKey`/`kind` are replay bookkeeping, deliberately excluded so the signature reflects the commands' user-visible effect, not compaction internals.
- `draft_command_log.args` is stored as `jsonb`, which does not preserve object key insertion order across a write/read round trip. `args` is canonicalized (object keys sorted recursively; array order preserved, since position is semantic there) before `JSON.stringify` + SHA-256, so structurally-equal args always hash identically regardless of storage round-trip.
- **One serializer, one call site pattern**: both `draftPublishReview` (review time) and `DraftController.publishDraft` (publish time, inside the transaction, over the reloaded log) import the same `computeLogSignature` function. There is no second implementation to drift out of sync — an unchanged log is guaranteed to signature-match byte for byte by construction.

## Screenshots

No UI change (passing an extra RPC argument, no visible UI difference).

## Verification

- `apps/server/src/draft-log-signature.test.ts` (new) — unit tests pin determinism: key-order independence (object and nested), array-order sensitivity, command-order sensitivity, value-change sensitivity, exclusion of `id`/`compactionKey`/`kind`, empty-log handling.
- `apps/server/src/draft-controller.test.ts` — added:
  - `publishes when expectedLogSignature matches the unchanged reloaded log`
  - `expectedCommandCount alone does NOT catch same-length content drift (the blind spot)` — proves the pre-existing guard's gap using a seeded create+delete-same-key+create-new compaction scenario (raw `draft_command_log` rows, since the real `cmd()` vocabulary never sets `compactionKey`/`kind`).
  - `rejects same-length content drift that expectedCommandCount alone would miss` — the same seeded scenario, now guarded by `expectedLogSignature`, correctly rejects.
  - `rejects publish when expectedLogSignature does not match the reloaded log, even with a correct count`
- `apps/server/src/functions/drafts.test.ts` — extended the existing end-to-end review→publish test to assert `logSignature` shape and round-trip it through the real `publishDraft` RPC; added `rejects publish when expectedLogSignature does not match what draftPublishReview returned (end-to-end)` covering both the negative case (wrong signature → 500, draft survives) and the positive case (review-produced signature → publishes cleanly).
- Ran the full server test suite locally: 380/380 passing. Lint clean on `apps/server`, `packages/app-data`, `packages/app`.
- Typecheck: pre-existing environment issue (dual `drizzle-orm` resolution between the root workspace and the `libs/wystack` git submodule's independent lockfile — confirmed present on `main` before this change, identical error count with and without this diff) prevents a clean `tsc` run repo-wide; not something this diff introduces or is in scope to fix. No draft-related typecheck errors appear in the diff.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR closes a blind spot in the draft publish drift guard: the existing `expectedCommandCount` check could be fooled by compaction, which can drop cancelled commands and let a different command backfill the count at the same length. It introduces a deterministic SHA-256 content signature (`computeLogSignature`) over `[{path, args}]` in replay order, computed server-side at review time and recomputed inside the publish transaction for comparison.

- **`draft-log-signature.ts`** (new): single-function canonicalizer that sorts object keys recursively (preserving array order), uses `Object.fromEntries` to handle `__proto__` keys correctly, and SHA-256 hashes the result. Comprehensive unit tests pin all determinism properties including the prototype-pollution edge case.
- **`DraftController.publishDraft`** / **`draftPublishReview`**: `expectedLogSignature` is enforced inside the same transaction as the count guard (both read the same reloaded log), and `draftPublishReview` returns `logSignature` for the client to ferry back verbatim.
- **Client surfaces** (`publish/page.tsx`, `DraftReviewPanel`): capture `logSignature` from the review response and pass it back on publish; no UI changes.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge. All guard checks run inside the publish transaction against the reloaded log, the canonicalizer is well-tested, and no existing behaviour is broken.

The change is narrowly scoped — a new server-side hash function plus plumbing to thread an opaque hex string from review to publish. The implementation is correct: same function at both call sites, canonicalization handles JSONB key-order and prototype-pollution edge cases, both guards read the same reloaded log atomically, and the test suite covers the blind spot it closes as well as all determinism properties. The only finding is a stale phrase in a JSDoc comment.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/server/src/draft-log-signature.ts | New file implementing SHA-256 content signature for the drift guard. `canonicalize` correctly uses `Object.fromEntries` (not bracket assignment) to handle `__proto__` keys, and sorts keys recursively while preserving array order. Unit-tested for determinism, key-order independence, and prototype-pollution edge cases. |
| apps/server/src/draft-controller.ts | Adds `expectedLogSignature` field to `PublishDraftOptions` and enforces it inside the publish transaction immediately after the count check. Both guards read the same reloaded log, have distinct error messages, and run before `beforeReplay`. Minor: the `beforeReplay` JSDoc first sentence omits `expectedLogSignature` from the guard enumeration. |
| apps/server/src/functions/drafts.ts | Adds `logSignature: computeLogSignature(commands)` to the `DraftPublishReview` response. Signature is computed over the full (unsummarized) `commands` array — correct, since `summarizeCommands` strips `args` before the response leaves the server. |
| apps/server/src/functions/draft-lifecycle.ts | Adds `expectedLogSignature` as an optional text RPC argument and forwards it verbatim to `publishDraft`. Correctly notes no parse step is needed (unlike the count, which is serialized to string for the RPC layer). |
| packages/app/src/components/assistant/DraftReviewPanel.tsx | Adds `logSignature` state, sets it from `review.logSignature` on successful review, and spreads it into the publish options when non-null. Correctly added to the `useCallback` dependency array. |
| apps/server/src/draft-log-signature.test.ts | Comprehensive unit tests pinning all determinism properties: key-order independence (flat and nested), array-order sensitivity, command-order sensitivity, value-change sensitivity, `id`/`compactionKey`/`kind` exclusion, empty-log, and the `__proto__` prototype-pollution edge case. |
| apps/server/src/draft-controller.test.ts | Adds four tests: happy-path signature match, proof-of-blind-spot (count guard alone passes on same-length drift), signature rejects same-length drift, and signature rejects with correct count. The `seedSameLengthContentDrift` helper correctly seeds raw rows and triggers compaction via an empty append. |
| apps/server/src/functions/drafts.test.ts | Extends the end-to-end review→publish test to verify the `logSignature` hex format and round-trip, and adds a rejection test proving a wrong signature returns 500 while leaving the draft intact. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client
    participant Server as Server (RPC)
    participant Controller as DraftController
    participant DB as draft_command_log

    Client->>Server: "draftPublishReview({ draftId })"
    Server->>DB: getDraftLog(draftId)
    DB-->>Server: commands[]
    Server->>Server: computeLogSignature(commands)
    Server-->>Client: "{ commandCount, logSignature, commands (redacted), diff }"

    Note over Client: Stores commandCount + logSignature in state

    Client->>Server: "publishDraft({ draftId, expectedCommandCount, expectedLogSignature })"
    Server->>Controller: publishDraft(draftId, options)

    rect rgb(230, 240, 255)
        Note over Controller,DB: Inside single transaction
        Controller->>DB: readLog(draftId, tx)
        DB-->>Controller: log[] (reloaded)
        Controller->>Controller: "log.length !== expectedCommandCount?"
        Note over Controller: throws (count mismatch)
        Controller->>Controller: "computeLogSignature(log) !== expectedLogSignature?"
        Note over Controller: throws (content drift)
        Controller->>Controller: assertPublishLogHasNoLateBound(log)
        Controller->>Controller: beforeReplay hook
        Controller->>DB: applyCommands (replay)
        Controller->>DB: deleteLog + sweepShadows
    end

    Controller-->>Server: CommitResult
    Server-->>Client: "{ tablesWritten }"
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client
    participant Server as Server (RPC)
    participant Controller as DraftController
    participant DB as draft_command_log

    Client->>Server: "draftPublishReview({ draftId })"
    Server->>DB: getDraftLog(draftId)
    DB-->>Server: commands[]
    Server->>Server: computeLogSignature(commands)
    Server-->>Client: "{ commandCount, logSignature, commands (redacted), diff }"

    Note over Client: Stores commandCount + logSignature in state

    Client->>Server: "publishDraft({ draftId, expectedCommandCount, expectedLogSignature })"
    Server->>Controller: publishDraft(draftId, options)

    rect rgb(230, 240, 255)
        Note over Controller,DB: Inside single transaction
        Controller->>DB: readLog(draftId, tx)
        DB-->>Controller: log[] (reloaded)
        Controller->>Controller: "log.length !== expectedCommandCount?"
        Note over Controller: throws (count mismatch)
        Controller->>Controller: "computeLogSignature(log) !== expectedLogSignature?"
        Note over Controller: throws (content drift)
        Controller->>Controller: assertPublishLogHasNoLateBound(log)
        Controller->>Controller: beforeReplay hook
        Controller->>DB: applyCommands (replay)
        Controller->>DB: deleteLog + sweepShadows
    end

    Controller-->>Server: CommitResult
    Server-->>Client: "{ tablesWritten }"
```

</a>
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(drafts): address review feedback on ..."](https://github.com/youhaowei/dashframe/commit/4ce4cb83deb4a6756746b680c09dfc45c6caa3bf) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41355264)</sub>

### CODE REVIEW

File: apps/server/src/draft-controller.ts [hunk 1]
```diff
@@ -476,15 +496,28 @@ export function createDraftController(
...
           log.length !== options.expectedCommandCount
         ) {
           throw new Error("publishDraft: draft changed since review");
         }
+        if (
+          options.expectedLogSignature !== undefined &&
+          computeLogSignature(log) !== options.expectedLogSignature
+        ) {
+          throw new Error("publishDraft: draft changed since review");
+        }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/draft-controller.ts:524]: Fixed in 4ce4cb83 — the count and signature mismatches now throw distinct messages ("...(count mismatch)" vs "...(content drift)"), so they're distinguishable in logs/stack traces. The client's substring match on "changed since review" still catches both.

File: apps/server/src/draft-log-signature.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ */
+function canonicalize(value: unknown): unknown {
+  if (Array.isArray(value)) {
+    return value.map(canonicalize);
+  }
+  if (value !== null && typeof value === "object") {
+    const sortedKeys = Object.keys(value as Record<string, unknown>).sort();
+    const out: Record<string, unknown> = {};
+    for (const key of sortedKeys) {
+      out[key] = canonicalize((value as Record<string, unknown>)[key]);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/draft-log-signature.ts]: Good catch, fixed in 4ce4cb83 — canonicalize now builds the sorted object via Object.fromEntries instead of out[key] = ..., which defines each entry as a genuine own property rather than tripping the __proto__ setter. Added a regression test (draft-log-signature.test.ts) that parses a __proto__-named key from JSON and asserts the two variants hash differently; confirmed it fails against the old assignment-based implementation.
